// @stacks/common/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BLOCKSTACK_DEFAULT_GAIA_HUB_URL = exports.PRIVATE_KEY_UNCOMPRESSED_LENGTH = exports.PRIVATE_KEY_COMPRESSED_LENGTH = exports.PeerNetworkID = exports.TransactionVersion = exports.ChainID = void 0;
var ChainID;
(function (ChainID) {
    ChainID[ChainID["Testnet"] = 2147483648] = "Testnet";
    ChainID[ChainID["Mainnet"] = 1] = "Mainnet";
})(ChainID || (exports.ChainID = ChainID = {}));
var TransactionVersion;
(function (TransactionVersion) {
    TransactionVersion[TransactionVersion["Mainnet"] = 0] = "Mainnet";
    TransactionVersion[TransactionVersion["Testnet"] = 128] = "Testnet";
})(TransactionVersion || (exports.TransactionVersion = TransactionVersion = {}));
var PeerNetworkID;
(function (PeerNetworkID) {
    PeerNetworkID[PeerNetworkID["Mainnet"] = 385875968] = "Mainnet";
    PeerNetworkID[PeerNetworkID["Testnet"] = 4278190080] = "Testnet";
})(PeerNetworkID || (exports.PeerNetworkID = PeerNetworkID = {}));
exports.PRIVATE_KEY_COMPRESSED_LENGTH = 33;
exports.PRIVATE_KEY_UNCOMPRESSED_LENGTH = 32;
exports.BLOCKSTACK_DEFAULT_GAIA_HUB_URL = 'https://hub.blockstack.org';
