// @stacks/transactions/dist/cl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserialize = exports.serialize = exports.tuple = exports.error = exports.ok = exports.some = exports.none = exports.bufferFromUtf8 = exports.bufferFromAscii = exports.bufferFromHex = exports.buffer = exports.stringUtf8 = exports.stringAscii = exports.list = exports.standardPrincipal = exports.contractPrincipal = exports.address = exports.principal = exports.uint = exports.int = exports.bool = exports.prettyPrint = void 0;
const common_1 = require("@stacks/common");
const clarity_1 = require("./clarity");
var prettyPrint_1 = require("./clarity/prettyPrint");
Object.defineProperty(exports, "prettyPrint", { enumerable: true, get: function () { return prettyPrint_1.prettyPrint; } });
exports.bool = clarity_1.boolCV;
exports.int = clarity_1.intCV;
exports.uint = clarity_1.uintCV;
function principal(address) {
    const [addr, name] = address.split('.');
    return name ? (0, clarity_1.contractPrincipalCV)(addr, name) : (0, clarity_1.standardPrincipalCV)(addr);
}
exports.principal = principal;
exports.address = principal;
exports.contractPrincipal = clarity_1.contractPrincipalCV;
exports.standardPrincipal = clarity_1.standardPrincipalCV;
exports.list = clarity_1.listCV;
exports.stringAscii = clarity_1.stringAsciiCV;
exports.stringUtf8 = clarity_1.stringUtf8CV;
exports.buffer = clarity_1.bufferCV;
const bufferFromHex = (hex) => (0, clarity_1.bufferCV)((0, common_1.hexToBytes)(hex));
exports.bufferFromHex = bufferFromHex;
const bufferFromAscii = (ascii) => (0, clarity_1.bufferCV)((0, common_1.asciiToBytes)(ascii));
exports.bufferFromAscii = bufferFromAscii;
const bufferFromUtf8 = (utf8) => (0, clarity_1.bufferCV)((0, common_1.utf8ToBytes)(utf8));
exports.bufferFromUtf8 = bufferFromUtf8;
exports.none = clarity_1.noneCV;
exports.some = clarity_1.someCV;
exports.ok = clarity_1.responseOkCV;
exports.error = clarity_1.responseErrorCV;
exports.tuple = clarity_1.tupleCV;
exports.serialize = clarity_1.serializeCV;
exports.deserialize = clarity_1.deserializeCV;
