// @stacks/transactions/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeTransaction = exports.StacksTransaction = exports.createSTXPostCondition = exports.createNonFungiblePostCondition = exports.createFungiblePostCondition = exports.Pc = exports.serializePayload = exports.isTokenTransferPayload = exports.isSmartContractPayload = exports.isPoisonPayload = exports.isContractCallPayload = exports.isCoinbasePayload = exports.Cl = exports.BytesReader = exports.isSingleSig = exports.emptyMessageSignature = void 0;
__exportStar(require("./authorization"), exports);
var authorization_1 = require("./authorization");
Object.defineProperty(exports, "emptyMessageSignature", { enumerable: true, get: function () { return authorization_1.emptyMessageSignature; } });
Object.defineProperty(exports, "isSingleSig", { enumerable: true, get: function () { return authorization_1.isSingleSig; } });
__exportStar(require("./builders"), exports);
var bytesReader_1 = require("./bytesReader");
Object.defineProperty(exports, "BytesReader", { enumerable: true, get: function () { return bytesReader_1.BytesReader; } });
exports.Cl = __importStar(require("./cl"));
__exportStar(require("./clarity"), exports);
__exportStar(require("./common"), exports);
__exportStar(require("./constants"), exports);
__exportStar(require("./contract-abi"), exports);
__exportStar(require("./keys"), exports);
var payload_1 = require("./payload");
Object.defineProperty(exports, "isCoinbasePayload", { enumerable: true, get: function () { return payload_1.isCoinbasePayload; } });
Object.defineProperty(exports, "isContractCallPayload", { enumerable: true, get: function () { return payload_1.isContractCallPayload; } });
Object.defineProperty(exports, "isPoisonPayload", { enumerable: true, get: function () { return payload_1.isPoisonPayload; } });
Object.defineProperty(exports, "isSmartContractPayload", { enumerable: true, get: function () { return payload_1.isSmartContractPayload; } });
Object.defineProperty(exports, "isTokenTransferPayload", { enumerable: true, get: function () { return payload_1.isTokenTransferPayload; } });
Object.defineProperty(exports, "serializePayload", { enumerable: true, get: function () { return payload_1.serializePayload; } });
exports.Pc = __importStar(require("./pc"));
var postcondition_1 = require("./postcondition");
Object.defineProperty(exports, "createFungiblePostCondition", { enumerable: true, get: function () { return postcondition_1.createFungiblePostCondition; } });
Object.defineProperty(exports, "createNonFungiblePostCondition", { enumerable: true, get: function () { return postcondition_1.createNonFungiblePostCondition; } });
Object.defineProperty(exports, "createSTXPostCondition", { enumerable: true, get: function () { return postcondition_1.createSTXPostCondition; } });
__exportStar(require("./postcondition-types"), exports);
__exportStar(require("./signature"), exports);
__exportStar(require("./signer"), exports);
__exportStar(require("./structuredDataSignature"), exports);
var transaction_1 = require("./transaction");
Object.defineProperty(exports, "StacksTransaction", { enumerable: true, get: function () { return transaction_1.StacksTransaction; } });
Object.defineProperty(exports, "deserializeTransaction", { enumerable: true, get: function () { return transaction_1.deserializeTransaction; } });
__exportStar(require("./types"), exports);
__exportStar(require("./utils"), exports);
