// @stacks/transactions/dist/clarity/prettyPrint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prettyPrint = void 0;
const common_1 = require("@stacks/common");
const _1 = require(".");
function formatSpace(space, depth, end = false) {
    if (!space)
        return ' ';
    return `\n${' '.repeat(space * (depth - (end ? 1 : 0)))}`;
}
function formatList(cv, space, depth = 1) {
    if (cv.list.length === 0)
        return '(list)';
    const spaceBefore = formatSpace(space, depth, false);
    const endSpace = space ? formatSpace(space, depth, true) : '';
    const items = cv.list.map(v => prettyPrintWithDepth(v, space, depth)).join(spaceBefore);
    return `(list${spaceBefore}${items}${endSpace})`;
}
function formatTuple(cv, space, depth = 1) {
    if (Object.keys(cv.data).length === 0)
        return '{}';
    const items = [];
    for (const [key, value] of Object.entries(cv.data)) {
        items.push(`${key}: ${prettyPrintWithDepth(value, space, depth)}`);
    }
    const spaceBefore = formatSpace(space, depth, false);
    const endSpace = formatSpace(space, depth, true);
    return `{${spaceBefore}${items.join(`,${spaceBefore}`)}${endSpace}}`;
}
function exhaustiveCheck(param) {
    throw new Error(`invalid clarity value type: ${param}`);
}
function prettyPrintWithDepth(cv, space = 0, depth) {
    if (cv.type === _1.ClarityType.BoolFalse)
        return 'false';
    if (cv.type === _1.ClarityType.BoolTrue)
        return 'true';
    if (cv.type === _1.ClarityType.Int)
        return cv.value.toString();
    if (cv.type === _1.ClarityType.UInt)
        return `u${cv.value.toString()}`;
    if (cv.type === _1.ClarityType.StringASCII)
        return `"${cv.data}"`;
    if (cv.type === _1.ClarityType.StringUTF8)
        return `u"${cv.data}"`;
    if (cv.type === _1.ClarityType.PrincipalContract)
        return `'${(0, _1.principalToString)(cv)}`;
    if (cv.type === _1.ClarityType.PrincipalStandard)
        return `'${(0, _1.principalToString)(cv)}`;
    if (cv.type === _1.ClarityType.Buffer)
        return `0x${(0, common_1.bytesToHex)(cv.buffer)}`;
    if (cv.type === _1.ClarityType.OptionalNone)
        return 'none';
    if (cv.type === _1.ClarityType.OptionalSome)
        return `(some ${prettyPrintWithDepth(cv.value, space, depth)})`;
    if (cv.type === _1.ClarityType.ResponseOk)
        return `(ok ${prettyPrintWithDepth(cv.value, space, depth)})`;
    if (cv.type === _1.ClarityType.ResponseErr)
        return `(err ${prettyPrintWithDepth(cv.value, space, depth)})`;
    if (cv.type === _1.ClarityType.List) {
        return formatList(cv, space, depth + 1);
    }
    if (cv.type === _1.ClarityType.Tuple) {
        return formatTuple(cv, space, depth + 1);
    }
    exhaustiveCheck(cv);
}
function prettyPrint(cv, space = 0) {
    return prettyPrintWithDepth(cv, space, 0);
}
exports.prettyPrint = prettyPrint;
