// @stacks/transactions/dist/clarity/serialize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeCV = void 0;
const common_1 = require("@stacks/common");
const types_1 = require("../types");
const postcondition_types_1 = require("../postcondition-types");
const constants_1 = require("./constants");
const errors_1 = require("../errors");
const constants_2 = require("../constants");
function bytesWithTypeID(typeId, bytes) {
    return (0, common_1.concatArray)([typeId, bytes]);
}
function serializeBoolCV(value) {
    return new Uint8Array([value.type]);
}
function serializeOptionalCV(cv) {
    if (cv.type === constants_1.ClarityType.OptionalNone) {
        return new Uint8Array([cv.type]);
    }
    else {
        return bytesWithTypeID(cv.type, serializeCV(cv.value));
    }
}
function serializeBufferCV(cv) {
    const length = new Uint8Array(4);
    (0, common_1.writeUInt32BE)(length, cv.buffer.length, 0);
    return bytesWithTypeID(cv.type, (0, common_1.concatBytes)(length, cv.buffer));
}
function serializeIntCV(cv) {
    const bytes = (0, common_1.bigIntToBytes)((0, common_1.toTwos)(cv.value, BigInt(constants_2.CLARITY_INT_SIZE)), constants_2.CLARITY_INT_BYTE_SIZE);
    return bytesWithTypeID(cv.type, bytes);
}
function serializeUIntCV(cv) {
    const bytes = (0, common_1.bigIntToBytes)(cv.value, constants_2.CLARITY_INT_BYTE_SIZE);
    return bytesWithTypeID(cv.type, bytes);
}
function serializeStandardPrincipalCV(cv) {
    return bytesWithTypeID(cv.type, (0, types_1.serializeAddress)(cv.address));
}
function serializeContractPrincipalCV(cv) {
    return bytesWithTypeID(cv.type, (0, common_1.concatBytes)((0, types_1.serializeAddress)(cv.address), (0, types_1.serializeLPString)(cv.contractName)));
}
function serializeResponseCV(cv) {
    return bytesWithTypeID(cv.type, serializeCV(cv.value));
}
function serializeListCV(cv) {
    const bytesArray = [];
    const length = new Uint8Array(4);
    (0, common_1.writeUInt32BE)(length, cv.list.length, 0);
    bytesArray.push(length);
    for (const value of cv.list) {
        const serializedValue = serializeCV(value);
        bytesArray.push(serializedValue);
    }
    return bytesWithTypeID(cv.type, (0, common_1.concatArray)(bytesArray));
}
function serializeTupleCV(cv) {
    const bytesArray = [];
    const length = new Uint8Array(4);
    (0, common_1.writeUInt32BE)(length, Object.keys(cv.data).length, 0);
    bytesArray.push(length);
    const lexicographicOrder = Object.keys(cv.data).sort((a, b) => a.localeCompare(b));
    for (const key of lexicographicOrder) {
        const nameWithLength = (0, postcondition_types_1.createLPString)(key);
        bytesArray.push((0, types_1.serializeLPString)(nameWithLength));
        const serializedValue = serializeCV(cv.data[key]);
        bytesArray.push(serializedValue);
    }
    return bytesWithTypeID(cv.type, (0, common_1.concatArray)(bytesArray));
}
function serializeStringCV(cv, encoding) {
    const bytesArray = [];
    const str = encoding == 'ascii' ? (0, common_1.asciiToBytes)(cv.data) : (0, common_1.utf8ToBytes)(cv.data);
    const len = new Uint8Array(4);
    (0, common_1.writeUInt32BE)(len, str.length, 0);
    bytesArray.push(len);
    bytesArray.push(str);
    return bytesWithTypeID(cv.type, (0, common_1.concatArray)(bytesArray));
}
function serializeStringAsciiCV(cv) {
    return serializeStringCV(cv, 'ascii');
}
function serializeStringUtf8CV(cv) {
    return serializeStringCV(cv, 'utf8');
}
function serializeCV(value) {
    switch (value.type) {
        case constants_1.ClarityType.BoolTrue:
        case constants_1.ClarityType.BoolFalse:
            return serializeBoolCV(value);
        case constants_1.ClarityType.OptionalNone:
        case constants_1.ClarityType.OptionalSome:
            return serializeOptionalCV(value);
        case constants_1.ClarityType.Buffer:
            return serializeBufferCV(value);
        case constants_1.ClarityType.UInt:
            return serializeUIntCV(value);
        case constants_1.ClarityType.Int:
            return serializeIntCV(value);
        case constants_1.ClarityType.PrincipalStandard:
            return serializeStandardPrincipalCV(value);
        case constants_1.ClarityType.PrincipalContract:
            return serializeContractPrincipalCV(value);
        case constants_1.ClarityType.ResponseOk:
        case constants_1.ClarityType.ResponseErr:
            return serializeResponseCV(value);
        case constants_1.ClarityType.List:
            return serializeListCV(value);
        case constants_1.ClarityType.Tuple:
            return serializeTupleCV(value);
        case constants_1.ClarityType.StringASCII:
            return serializeStringAsciiCV(value);
        case constants_1.ClarityType.StringUTF8:
            return serializeStringUtf8CV(value);
        default:
            throw new errors_1.SerializationError('Unable to serialize. Invalid Clarity Value.');
    }
}
exports.serializeCV = serializeCV;
